File: AmpedBiz/AmpedBiz.Web/src/common/models/purchase-order-event.ts
import {ProductInventory} from './product';
import {Lookup} from '../custom_types/lookup';
import {Dictionary} from '../custom_types/dictionary';
import {PurchaseOrder, PurchaseOrderStatus, PurchaseOrderItem, PurchaseOrderPayment, PurchaseOrderReceipt} from './purchase-order';


export interface PurchaseOrderEvent {
  id?: string;
  purchaseOrderId?: string;
  transitionDescription?: string;
}

export interface PurchaseOrderNewlyCreatedEvent extends PurchaseOrderEvent {
  createdBy?: Lookup<string>;
  createdOn?: Date;
  expectedOn?: Date;
  paymentType?: Lookup<string>;
  shipper?: Lookup<string>;
  shippingFeeAmount?: number;
  taxAmount?: number;
  supplier?: Lookup<string>;
  items?: PurchaseOrderItem[];
}

export interface PurchaseOrderSubmittedEvent extends PurchaseOrderEvent {
  submittedBy?: Lookup<string>;
  submittedOn?: Date;
}

export interface PurchaseOrderApprovedEvent extends PurchaseOrderEvent {
  approvedBy?: Lookup<string>;
  approvedOn?: Date;
}

export interface PurchaseOrderPaidEvent extends PurchaseOrderEvent {
  payments?: PurchaseOrderPayment[],
}

export interface PurchaseOrderReceivedEvent extends PurchaseOrderEvent {
  receipts?: PurchaseOrderReceipt[];
}

export interface PurchaseOrderCompletedEvent extends PurchaseOrderEvent {
  completedBy?: Lookup<string>;
  completedOn?: Date;
}

export interface PurchaseOrderCancelledEvent extends PurchaseOrderEvent {
  cancelledBy?: Lookup<string>;
  cancelledOn?: Date;
  cancellationReason?: string;
}

File: AmpedBiz/AmpedBiz.Service/Dto/PurchaseOrderEvent.cs
using AmpedBiz.Common.CustomTypes;
using System;
using System.Collections.Generic;

namespace AmpedBiz.Service.Dto
{
    public class PurchaseOrderEvent
    {
        public Guid Id { get; set; }

        public Guid PurchaseOrderId { get; set; }

        public string TransitionDescription { get; set; }
    }

    public class PurchaseOrderNewlyCreatedEvent : PurchaseOrderEvent
    {
        public Lookup<Guid> CreatedBy { get; set; }

        public DateTime? CreatedOn { get; set; }

        public DateTime? ExpectedOn { get; set; }

        public Lookup<string> PaymentType { get; set; }

        public Lookup<string> Shipper { get; set; }

        public Lookup<string> Supplier { get; set; }

        public decimal ShippingFeeAmount { get; set; }

        public decimal TaxAmount { get; set; }

        public IEnumerable<PurchaseOrderItem> Items { get; set; }
    }

    public class PurchaseOrderSubmittedEvent : PurchaseOrderEvent
    {
        public Lookup<Guid> SubmittedBy { get; set; }

        public DateTime? SubmittedOn { get; set; }
    }

    public class PurchaseOrderApprovedEvent : PurchaseOrderEvent
    {
        public Lookup<Guid> ApprovedBy { get; set; }

        public DateTime? ApprovedOn { get; set; }
    }

    public class PurchaseOrderPaidEvent : PurchaseOrderEvent
    {
        public IEnumerable<PurchaseOrderPayment> Payments { get; set; }
    }

    public class PurchaseOrderReceivedEvent : PurchaseOrderEvent
    {
        public IEnumerable<PurchaseOrderReceipt> Receipts { get; set; }
    }

    public class PurchaseOrderCompletedEvent : PurchaseOrderEvent
    {
        public Lookup<Guid> CompletedBy { get; set; }

        public DateTime? CompletedOn { get; set; }
    }

    public class PurchaseOrderCancelledEvent : PurchaseOrderEvent
    {
        public Lookup<Guid> CancelledBy { get; set; }

        public DateTime? CancelledOn { get; set; }

        public string CancellationReason { get; set; }
    }
}
